Coach Quiz Exercise Resources

nucleogenesis opened the pull request:
<!--
 1. Following guidance below, replace …'s with your own words
 2. After saving the PR, tick of completed checklist items
 3. Skip checklist items that are not applicable or not necessary
 4. Delete instruction/comment blocks
-->

## Summary
<!--
 * description of the change
 * manual verification steps performed
 * screenshots if the PR affects the UI
-->

- Introduces new `useExerciseResources` module
- Implements basic interface for fetching topics & exercises, annotated with their respective number of assessment descendants
- Hooks that up into previous work done by @AllanOXDi on the Select Resources page so that the user can navigate the tree

#### Outstanding issues

Some or all of these may not be entirely necessary to implement in this PR, but I've been trying to work out an answer to some of these issues with varying degrees of success:

**The router situation is a bit screwy**

As it is now the component in which we're intending to render and update the list of content cards is being reloaded/remounted every time the route changes.

The intention for this is that whenever the route changes (specifically, the `:topic_id` param) then we'll be able to push that onto the state and the browser back/forward navigation functions will allow the user to walk back and forth through the tree as one would expect.

I've played with and tried a few solutions that have tried to sort of tweak what is already here but none of them have been effective.

I've also:
- Ensured there is no `:key` on any router-view which would force re-render
- Tried using named router-views w/ the nested routes (although I feel that maybe I missed something in this implementation w/ the structure of the routes)
- Tried using `<keep-alive>` on the router-view in CoachImmersivePage (and the others I tried to use as well)

The result here is that the `useExerciseResources` module is re-initialized every time the route changes which is _Extremely Not Good ™️_ for the user or for the developer using the module and relying on it to just do the magic it ought to, update it's refs and allow the components to derive their UI from that state.

**ContentNodeResource.fetchTree and the figuring out of checkboxes**

The designs expect that we allow checkboxes on topics and I think that this is the ideal solution, with some exceptions.

When I spoke w/ @rtibbles about this we discussed that basically we should just not allow the checking of a topic folder unless we can get all of the possible assessments from within the currently-fetched tree (which is limited to two generations). However, the child content nodes themselves are not annotated with `num_assessments` which makes figuring out whether or not we can select the topic and all of its children difficult.

I think maybe this would be worth a co-hack at some point and I don't consider solving this to be a blocking issue for this issue as it would be a fine follow-up issue.

**Clean-up**

- [ ] Remove all `channels` and `bookmarks` business from `useQuizCreation` and anywhere it is used, replace it with those from `useExerciseResources` (it may not be referenced anywhere else though).
- [ ] Clean-up the coach plan constants' `PageNames` map, removing the `PageNames` which we are no longer using

## Reviewer Guidance

- Please look at the `useExerciseResources` in particular. It only currently handles the basic fetching-and-annotating.
- Does the overarching structure for the module seem reasonable (see the comments within the module)?

Any and all thoughts & suggestions around the router would be greatly appreciated.

nucleogenesis committed:
useExerciseResources foundations

nucleogenesis committed:
fetch and annotate resources in useExerciseResources

nucleogenesis committed:
apply changes from previous work

rtibbles commented on kolibri/plugins/coach/assets/src/composables/useExerciseResources.js:
```diff
@@ -0,0 +1,213 @@
+  /** @type {ref<Boolean>} Whether we are currently fetching/processing the child nodes */
+  const _loading = ref(false);
+
+  const _kindsToFetch = ref([ContentNodeKinds.EXERCISE, ContentNodeKinds.TOPIC]);
```
It seems like this is never modified, and rather seems to be a constant - so no real need for this to be a ref.

rtibbles commented on kolibri/plugins/coach/assets/src/composables/useExerciseResources.js:
```diff
@@ -0,0 +1,213 @@
+ */
+export default function useExerciseResources() {
+  /** @type {ref<string>} The topicId for which we will reflect the child nodes */
+  const _currentTopicId = ref(null);
```
I think we could simplify some of the logic here by accepting slight coupling between the VueRouter route and this module. It would be relatively simple to turn this into a computed property that reads directly from the route state, see here for an example: https://github.com/learningequality/kolibri/blob/release-v0.16.x/kolibri/plugins/learn/assets/src/composables/useSearch.js#L153

(in future versions, VueRouter directly exposes composable state, so we could migrate to using that).

rtibbles commented on kolibri/plugins/coach/assets/src/composables/useExerciseResources.js:
```diff
@@ -0,0 +1,213 @@
+        'You should be passing a topic that you found in the contentList.'
+      );
+    }
+    if (get(_contentList).children) {
```
This suggests that `_contentList` might actually be a ref pointing to a topic that has `children` rather than to an array of content nodes? Feels like the ability to be either might be confusing.

rtibbles commented on kolibri/plugins/coach/assets/src/composables/useExerciseResources.js:
```diff
@@ -0,0 +1,213 @@
+      );
+    }
+    if (get(_contentList).children) {
+      const childTopicIds = get(_contentList)
```
So, if the `_contentList` ref value has the attribute `children`, then we should treat it as an array? Something very odd is happening here.

rtibbles commented on kolibri/plugins/coach/assets/src/composables/useExerciseResources.js:
```diff
@@ -0,0 +1,213 @@
+
+  function fetchMore() {
+    if (get(_moreParams)) {
+      fetchContentList(_moreParams);
```
This would need to be wrapped in `get` as well.

rtibbles commented on kolibri/plugins/coach/assets/src/composables/useExerciseResources.js:
```diff
@@ -0,0 +1,213 @@
+
+    // Set the default params for the fetchTree method .
+    const params = {
+      ...(moreParams || {}),
```
Not that the `more` object from a previous paginated fetch will include the previously used query parameters, so adding both is slightly redundant.

It might be good to separate out the middle part of this function with the behaviour that is common to 'base' fetching and 'more' fetching into a helper function, and then create `fetchContentList` and `fetchMoreContent` functions that do the specific parameter injection, and result processing.

rtibbles commented on kolibri/plugins/coach/assets/src/composables/useExerciseResources.js:
```diff
@@ -0,0 +1,213 @@
+  /**
+   * @param {string} topicId The topicId for which we want to fetch the descendants
+   */
+  function setCurrentTopicId(topicId) {
```
I think the logic here could be handled by directly watching the computed topicId I suggested above.

rtibbles commented on kolibri/plugins/coach/assets/src/composables/useExerciseResources.js:
```diff
@@ -0,0 +1,213 @@
+    console.log('initializing resources');
+    set(_loading, true);
+    Promise.all([fetchChannelResources(), fetchBookmarks()])
+      .then(() => setCurrentTopicId(null))
```
This line is currently causing a race condition when the page is first loaded - if this gets set after the topic id gets set from the route, then this will cause weird rendering.

With the computed prop for currentTopicId based on the route that I suggest above, this line can be removed, and will help prevent this problem.

rtibbles commented on kolibri/plugins/coach/assets/src/routes/planExamRoutes.js:
```diff
@@ -49,6 +49,23 @@ export default [
         name: PageNames.QUIZ_SELECT_RESOURCES,
         path: ':section_id/select-resources',
+
+        children: [
```
As a general rule, nested routes are a bad idea - I've tried them, and yet to find a particularly good reason for using them.

The main place they might have value is when you need to swap out different components within another view, as per the example in the docs: https://v3.router.vuejs.org/guide/essentials/nested-routes.html

However, that's not what you're doing here, where it's just being used to gate between the base route (show channels) and a specific topic route (show a topic). This can be better achieved by just using an optional URL parameter for topic_id in the route above. @AllanOXDi has implemented this already, I believe.

rtibbles reviewed (commented):
A few notes on implementation - some cleanup that I think can help consistency of behaviour here.

nucleogenesis committed:
Flatten createExamRoutes

Co-authored-by: Allan Otodi Opeto <otodi.allan1@gmail.com>

nucleogenesis committed:
refactoring fetching and annotation of contents

splits the business of fetching (and fetching more) in the fetchTree
method into its own module

then another new module repurposes code from useExerciseResources and
leverages the new useFetchTree module to fetch the contents and
annotate them as needed for navigation the content tree specifically for
use with quizzes

useFetchTree (at least tries) to straightforwardly provide the intial
fetch of content while keeping itself up-to-date whenever the user
fetches more

rtibbles commented on kolibri/plugins/coach/assets/src/views/plan/CreateExamPage/ResourceSelection.vue:
```diff
@@ -132,31 +139,95 @@
+        return _loading.value || quizResourcesLoading.value;
+      });
+
+      onMounted(() => {
```
I don't think this needs to be wrapped in an `onMounted` hook - nothing in here is interacting with the DOM, so this would be called in the equivalent of the `created` life cycle hook, which is just in the `setup` function.

Should also watch the `topicId` to reload in the case it changes within the component's lifecycle (which it will once we've sorted out the global hook issues).

rtibbles commented on kolibri/plugins/coach/assets/src/views/plan/CreateExamPage/ResourceSelection.vue:
```diff
@@ -132,31 +139,95 @@
     mixins: [commonCoreStrings],
     setup() {
+      const store = getCurrentInstance().proxy.$store;
+      const topicId = store.state.route.params.topic_id;
```
I think you'll need to wrap this in `computed(() => store.state.route.params.topic_id)` for it to be reactive, and hence watchable later.

rtibbles commented on kolibri/plugins/coach/assets/src/composables/useFetchTree.js:
```diff
@@ -0,0 +1,78 @@
+   **/
+  function fetchTree(paramOverrides = get(_moreParams) || {}) {
+    set(_loading, true);
+    Object.assign(params, paramOverrides);
```
This is mutating the `params` object that is passed into the initialization of the composable, which feels like you may be contaminating the composable's state. This may not have an effect, but this feels like it could have unintended consequences.

The main possibility that comes to mind is if the topicId changes, and we call `fetchTree` again, but we still have the `_moreParams` from the previous topicId - that would cause a right old mess.

I still think it would be better to have a separate implementation for fetchTree and fetchMore - the difference being whether you pass in `params` or `get(_moreParams)` - also, worth noting that in the case of `fetchMore` we should do nothing in the case that `get(_moreParams)` evaluates to null.

So I'd recommend changing this to a `_fetchNodeTree` or somesuch, then making:
```
function fetchTree() {
   return _fetchNodeTree(params);
}
```

```
function fetchMore() {
  const moreParams = get(_moreParams);
  if (!moreParams) {
    return Promise.reject('Tried to call fetchMore when no more ContentNodes are available);
  }
  return _fetchNodeTree(moreParams);
}
```

rtibbles commented on kolibri/plugins/coach/assets/src/composables/useFetchTree.js:
```diff
@@ -0,0 +1,78 @@
+    set(_loading, true);
+    Object.assign(params, paramOverrides);
+
+    return ContentNodeResource.fetchTree({ id: topicId, params }).then(topicTree => {
```
If we want to make this work across navigation events when the composable is not reinitialized, then the passed in `topicId` would have to be reactive, and so this would be `get(topicId)`.

rtibbles reviewed (changes requested):
Would still like to make this work for a reactive `topicId`, so that we can reliably use this when the entire component hierarchy is not being torn down on every navigation event.

nucleogenesis committed:
topicId is now passed around as a computedRef

nucleogenesis committed:
navigation works; channels & topics annotated, displayed appropriately

nucleogenesis committed:
view more button wired up - params are set correctly on fetchTree call

nucleogenesis committed:
fix issue where channels always showed on reload, unneeded "view more" btn

nucleogenesis committed:
cleanup - remove unused module, lint

nucleogenesis committed:
watch topicId and call fetch for when we fix the routing

rtibbles commented on kolibri/plugins/coach/assets/src/composables/useFetchTree.js:
```diff
@@ -0,0 +1,89 @@
+
+/**
+ * @deftype FetchTreeConfig
+ * @property {computedRef<string|null|undefined>} topicId - The id of the root node to fetch the
```
Technically, this could just be any reactive value - I'm not sure what the type for that is though :)

rtibbles commented on kolibri/plugins/coach/assets/src/composables/useFetchTree.js:
```diff
@@ -0,0 +1,89 @@
+ * @param
+ */
+export default function useFetchTree({ topicId, params = {} } = {}) {
+  const _topic = ref(null);
```
I hadn't asked before, but I am assuming the intention of the private refs combined with public computed props is to make the internal state only mutable by the composable itself - if so, would be good to document that as intentional as comments inline (I don't think it belongs in the function description).

nucleogenesis replied:
I've added a note of guidance above these re: the intention -- but I wonder if maybe putting `@private` in the JSDocs would serve the same effect well -- that is to say I wonder if that would be clear enough indication of the fact that they're only mutable locally in this context.

rtibbles commented on kolibri/plugins/coach/assets/src/composables/useFetchTree.js:
```diff
@@ -0,0 +1,89 @@
+    topic: computed(() => get(_topic)),
+    resources: computed(() => get(_resources)),
+    loading: computed(() => get(_loading)),
+    hasMore: computed(() => get(_moreParams) !== null),
```
Could define this further up in the composable, and use a check against the `hasMore` computed prop in the `fetchMore` method.

rtibbles commented on kolibri/plugins/coach/assets/src/composables/useQuizResources.js:
```diff
@@ -0,0 +1,138 @@
+
+/**
+ * @typedef {Object} QuizResourcesConfig
+ * @property { computed <string|null|undefined> } topicId - The id of the root node to fetch the
```
You used `computed` here, but `computedRef` above for the topicId - good to be consistent in our annotations!

nucleogenesis replied:
Yeah I think that @bjester mentioned during a previous code review that the proper "type" for it is called a "ComputedRef" but I ended up getting littered with errors on my JSDocs because it was not a found name so I reverted to using the functions that are imported in the file a la `ref` or `computed`, which appeased the LSP.

I'd definitely be interested to know if others' IDEs complain differently based on this.

bjester replied:
I think `ComputedRef` [is appropriate here](https://vuejs.org/guide/typescript/composition-api#typing-computed), and my IDE recognizes it properly. Although, it recognizes it in the sense that it does a weak match to (a) possible object(s) with the same name. I think `computed` would be incorrect because the value isn't an instance of `computed`. I can also understand your LSP not liking it if that name isn't in scope, within that file. I think having to import types starts leaning into TypeScript. I'm unsure what the best solution would be

rtibbles commented on kolibri/plugins/coach/assets/src/composables/useQuizResources.js:
```diff
@@ -0,0 +1,138 @@
+  async function annotateTopicsWithDescendantCounts(topicIds = []) {
+    return ContentNodeResource.fetchDescendantsAssessments(topicIds)
+      .then(({ data: topicsWithDescendantCounts }) => {
+        //console.log(topicsWithDescendantCounts);
```
delete

rtibbles commented on kolibri/plugins/coach/assets/src/composables/useQuizResources.js:
```diff
@@ -0,0 +1,138 @@
+          .filter(node => {
+            // Only keep topics which have assessments in them to begin with
+            if (node.kind === ContentNodeKinds.TOPIC) {
+              if (node.num_assessments > 0) {
```
Could reduce nesting slightly here with an early return strategy, and by just returning the value of this comparison:
```
if (node.kind !== ContentNodeKinds.TOPIC) {
  return true;
}
return node.num_assessments > 0;
```

rtibbles commented on kolibri/plugins/coach/assets/src/composables/useQuizResources.js:
```diff
@@ -0,0 +1,138 @@
+      .catch(e => {
+        // TODO Work out best UX for this situation -- it may depend on if we're fetching more
+        // or the initial list of contents
+        console.error(e);
```
Flagging that we've just merged no-console linting into 0.16, so it will be coming into develop soon.

If you wish to keep this console.error, please use `kolibri.lib.logging` to log the error instead. This has the distinct advantage of being turned off during tests.

rtibbles commented on kolibri/plugins/coach/assets/src/composables/useQuizResources.js:
```diff
@@ -0,0 +1,138 @@
+  async function fetchMoreQuizResources() {
+    set(_loading, true);
+    return fetchMore().then(async results => {
+      console.log(results);
```
delete!

rtibbles commented on kolibri/plugins/coach/assets/src/composables/useQuizResources.js:
```diff
@@ -0,0 +1,138 @@
+  }
+
+  /** @returns {Boolean} Whether the given node should be displayed with a checkbox
+   * currently passes when the node is an Exercise or if it is a topic with fewer than 20
```
Do we have explanatory text for this anywhere? (if so, probably good to store the value as a constant to feed into the text, and here, in case we decide to change it).

nucleogenesis replied:
I'm not 100% sure how the question references the highlighted lines.

I removed the condition around topic selection altogether so for now we'll only deal w/ allowing Exercises to be selected.

Given the complexity in identifying whether or not we have all of the topics descendants loaded by way of fetchTree -- namely with consideration for the fact that we may not have all of the first level's children if we `hasMore` -- this seems like the best way to go.

rtibbles replied:
Hrm, that's a little bit of a step back - as it can be helpful to select an entire topic full of exercises to create a quiz.

We had previously discussed only allowing topics that had no topic descendants to be selected - but I guess knowledge of that does slightly depend on whether you have fetched the grandchildren of the current topic or not.

rtibbles replied:
We discussed this on slack and came up with:

> the limit is you can select a folder if we fetched every child in one go (an implicit size limit of 12) and every child is not a folder

Will be filed as a follow up issue.

rtibbles commented on kolibri/plugins/coach/assets/src/constants/index.js:
```diff
@@ -16,6 +16,9 @@ export const PageNames = {
   QUIZ_REPLACE_QUESTIONS: 'QUIZ_REPLACE_QUESTIONS',
   QUIZ_SELECT_RESOURCES: 'QUIZ_SELECT_RESOURCES',
+  SELECT_FROM_RESOURCE: 'SELECT_FROM_RESOURCE',
+  BOOK_MARKED_RESOURCES: 'BOOK_MARKED_RESOURCES',
```
Of these three, I am only seeing this new PageName being used in the routes below?

rtibbles commented on kolibri/plugins/coach/assets/src/routes/planExamRoutes.js:
```diff
@@ -37,20 +37,26 @@ export default [
+  },
+  {
+    name: PageNames.BOOK_MARKED_RESOURCES,
+    path: '/:classId/plan/quizzes/new/:section_id/select-resources/bookmarks/:topic_id?',
```
I was curious, what does topic_id mean in this context? We can't bookmark topics, might this be more semantically named?

nucleogenesis replied:
I meant to remove this when it clicked in my head that we cannot bookmark topics - fixed it

rtibbles commented on kolibri/plugins/coach/assets/src/views/plan/CreateExamPage/ResourceSelection.vue:
```diff
@@ -1,20 +1,521 @@
+        </div>
+      </div>
+
+      <!-- <LessonsSearchFilters
```
Is this intended to be integrated? Presumably it would only be keyword search at the moment?

nucleogenesis replied:
Initially I suspected we might keep some of the search-related logic in place, but as we will be revamping the coach search experience altogether to make use of existing search business from Learn, I'll just remove old search stuff altogether. It can just be referenced by looking at a <=release-v0.16.x branch in the future if needed.

rtibbles replied:
Discussed on slack - we'll file a follow up issue to restore keyword search on top of this implementation, but we'll make it non-blocking for release.

rtibbles commented on kolibri/plugins/coach/assets/src/views/plan/CreateExamPage/ResourceSelection.vue:
```diff
@@ -1,20 +1,521 @@
+      // TODO let's not use text for this
+      const viewMoreButtonState = computed(() => {
+        if (hasMore.value) {
+          return 'yes';
```
Should these be constants?

nucleogenesis replied:
`return 'yes'` -- but we're using what was defined in a borrowed Lessons component, I'll track this separately as I suspect it'll require some updates to Lesson components using the same component as well.

rtibbles commented on kolibri/plugins/coach/assets/src/views/plan/CreateExamPage/ResourceSelection.vue:
```diff
@@ -1,20 +1,521 @@
+
+      // Load up the channels
+      if (!topicId.value) {
+        console.log('NO TOPIC ID');
```
delete

rtibbles commented on kolibri/plugins/coach/assets/src/views/plan/CreateExamPage/ResourceSelection.vue:
```diff
@@ -1,20 +1,521 @@
+          }),
+          ContentNodeResource.fetchBookmarks({ params: { limit: 25, available: true } }).then(
+            data => {
+              console.log('Bookmarks API results', data); // Do we have a `more` here?
```
delete

rtibbles commented on kolibri/plugins/coach/assets/src/views/plan/CreateExamPage/ResourceSelection.vue:
```diff
@@ -1,20 +1,521 @@
+      },
+      selectionMetadata(/*content*/) {
+        return function() {};
+        // let count = 0;
```
What's the TODO here?

rtibbles replied:
This seems to be covered in #11741

rtibbles reviewed (commented):
I think all my previous comments have been addressed, mostly questions about the code and next steps, and a little cleanup.

nucleogenesis committed:
fixes galore

- use kolibri.lib.logging in place of console.*
- remove unneeded console.logs
- add guidance re: private variables
- remove unused search-related business
- fix bookmarks route (no topic_id)
- remove unused pagename constants
- code cleanup thanks to @rtibbles suggestions
- improved comments
- indicated follow-up issues where commented-out TODO work remains

rtibbles commented on kolibri/plugins/coach/assets/src/views/plan/CreateExamPage/ResourceSelection.vue:
```diff
@@ -1,20 +1,425 @@
+
+        <p>{{ selectFromBookmarks$() }}</p>
+
+        <div @click="lessonCardClicked">
```
This click handler should be cleaned up - it seems like it's not needed anyway.

nucleogenesis committed:
remove unused handler

rtibbles reviewed (approved):
All code comments addressed - manual testing shows no errors apart from 'remains to be implemented'!

rtibbles commented:
Docs build has been fixed independently.
